For #17785 - Use screenshots setting when adding secure flag

ekager opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

sv-ohorvath commented:
Please rebase, the blocking test was disabled: https://github.com/mozilla-mobile/fenix/commit/f0b7b3b987202339a5affbe76c5b857405f004a6

Mugurell commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayDialogFragment.kt:
```diff
@@ -86,7 +86,11 @@ class TabTrayDialogFragment : AppCompatDialogFragment(), UserInteractionHandler
 
     private val collectionStorageObserver = object : TabCollectionStorage.Observer {
-        override fun onCollectionCreated(title: String, sessions: List<TabSessionState>, id: Long?) {
+        override fun onCollectionCreated(
```
nit: this new formatting will skew git history and make it harder when using blame to find why some changes were introduced.

Mugurell reviewed (approved):
Tested it and saw it resolving the reported issue.
How, when, where we actually set the `FLAG_SECURE` will probably need a bit more work for which I've created https://github.com/mozilla-mobile/fenix/issues/17854 and https://github.com/mozilla-mobile/fenix/issues/17855

ekager commented:
@Mugurell you're right! I decided to try to address that here because what we were doing before was not very robust. 

In my mind there are 3 times we should worry about setting this flag and none of them involve the Android lifecycle: 
1. When normal/private mode changes
2. When screenshot setting changes
3. When normal/private tab in tabs tray changes

I've addressed those 3 places now and removed the old lifecycle triggers, can you think of any other times we would care about this?

Mugurell commented on app/src/main/java/org/mozilla/fenix/HomeActivity.kt:
```diff
@@ -341,10 +341,6 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         settings().shouldReturnToBrowser =
             components.core.store.state.getNormalOrPrivateTabs(private = false).isNotEmpty()
 
-        if (settings().lastKnownMode.isPrivate) {
```
I think we still need this because of 
https://github.com/mozilla-mobile/fenix/issues/11153
Even if screenshots are allowed
We should do our best to not display the private content in the recents screen.

Mugurell replied:
If we keep this a counterpart in onResume will probably be needed.

ekager replied:
Oops totally forgot about this behavior. Good call

ekager replied:
Added comments explaining this too to avoid regressions.

Mugurell commented on app/src/main/java/org/mozilla/fenix/HomeActivity.kt:
```diff
@@ -850,7 +846,18 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
 
     protected open fun createBrowsingModeManager(initialMode: BrowsingMode): BrowsingModeManager {
         return DefaultBrowsingModeManager(initialMode, components.settings) { newMode ->
+            updateSecureWindowFlags(newMode)
```
👍

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -407,12 +406,6 @@ class HomeFragment : Fragment() {
             }
         }
 
-        if (browsingModeManager.mode.isPrivate) {
```
Tested, changing from PWA to private tabs works nicely.

Mugurell commented on app/src/main/java/org/mozilla/fenix/settings/PrivateBrowsingFragment.kt:
```diff
@@ -44,7 +46,18 @@ class PrivateBrowsingFragment : PreferenceFragmentCompat() {
         requirePreference<SwitchPreference>(R.string.pref_key_allow_screenshots_in_private_mode).apply {
-            onPreferenceChangeListener = SharedPreferenceUpdater()
+            onPreferenceChangeListener = object : SharedPreferenceUpdater() {
+                override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
```
👍

Mugurell reviewed (approved):
Played a bit with the app, everything seems to work nicely now! 
Other than having to support that edgecase with the recents screen I don't see other issues.

ekager commented:
Great! I'll merge once everything passes. Thanks for filing those issues - definitely good to nail down the correct behavior while we're already looking at this.

Mugurell commented:
@ekager I saw one more case that will probably need a special treatment to know that everything will be finished

Here, in onResume the SECURE flag is set but then it is not cleared (if not needed) in onPause.
https://github.com/mozilla-mobile/fenix/blob/07748f69b75144e8e247b742f66a756f999cb299/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/SavedLoginsFragment.kt#L60-L63
